midterm-2023-11-07/code/5130-stefan-vitanov/script.ts

```diff
@@ -0,0 +1,349 @@
+    directorDiv.textContent = movie.director;
+ 
+    const yearDiv = document.createElement('div');
+    yearDiv.classList.add('movie-data');
+    yearDiv.textContent = movie.year.toString();
+ 
+    const genreDiv = document.createElement('div');
+    genreDiv.classList.add('movie-data');
+    genreDiv.textContent = formatGenres(movie.genre);
+ 
+    const plotDiv = document.createElement('div');
+    plotDiv.classList.add('movie-data');
+    plotDiv.textContent = truncatePlot(movie.plot, 50);
+ 
+    const castDiv = document.createElement('div');
+    castDiv.classList.add('movie-data');
+    castDiv.innerHTML = formatCast(movie.cast);
+ 
+    const oscarsDiv = document.createElement('div');
+    oscarsDiv.classList.add('movie-data');
+    oscarsDiv.innerHTML = formatOscars(movie.oscars);
+ 
+ 
+    movieContainer.appendChild(idDiv);
+    movieContainer.appendChild(titleDiv);
+    movieContainer.appendChild(directorDiv);
+    movieContainer.appendChild(yearDiv);
+    movieContainer.appendChild(genreDiv);
+    movieContainer.appendChild(plotDiv);
+    movieContainer.appendChild(castDiv);
+    movieContainer.appendChild(oscarsDiv);
+ 
+    return movieContainer;
+  }
+ 
+ 
+  let currentSort: { field: keyof Movie | null, ascending: boolean } = {
+    field: null,
+    ascending: true
+  };
+ 
+  function sortMovies(movies: Movie[], sortField: keyof Movie, ascending: boolean): Movie[] {
+    return movies.sort((a, b) => {
+      let valA = a[sortField];
+      let valB = b[sortField];
+ 
+      if (sortField === 'genre') {
+ 
+        const genreLengthComparison = a.genre.length - b.genre.length;
+        if (genreLengthComparison !== 0) return ascending ? genreLengthComparison : -genreLengthComparison;
+ 
+ 
+        const sortedGenresA = [...a.genre].sort();
+        const sortedGenresB = [...b.genre].sort();
+        for (let i = 0; i < sortedGenresA.length; i++) {
+          const comparison = sortedGenresA[i].localeCompare(sortedGenresB[i]);
+          if (comparison !== 0) return ascending ? comparison : -comparison;
+        }
+ 
+        return 0; 
+      } else if (sortField === 'cast') {
+ 
+        const castLengthComparison = a.cast.length - b.cast.length;
+        if (castLengthComparison !== 0) return ascending ? castLengthComparison : -castLengthComparison;
+ 
+ 
+        for (let i = 0; i < a.cast.length; i++) {
+          const comparison = a.cast[i].actor.localeCompare(b.cast[i].actor);
+          if (comparison !== 0) return ascending ? comparison : -comparison;
+        }
+ 
+        return 0; 
+      } else if (typeof valA === 'string' && typeof valB === 'string') {
+ 
+        return ascending ? valA.localeCompare(valB) : valB.localeCompare(valA);
+      } else if (typeof valA === 'number' && typeof valB === 'number') {
+ 
+        return ascending ? valA - valB : valB - valA;
+      }
+ 
+ 
+      return 0;
+    });
+  }
+ 
+  async function populateGenreDropdown() {
+    const movies = await getMovies();
```

Comment:
The main issue in this code is that it's loading the data for each and every usage. This is wasteful, slows everything down, and for an actual API will incur extra costs.

An alternative approach would be to have `getMovies` to cache the data, or to keep the data in a state variable.